Mail2Bug/WorkItemManagement/TFSWorkItemManager.cs

```diff
@@ -251,7 +253,66 @@ public void AttachFiles(int workItemId, List<string> fileList)
                 WorkItem workItem = _tfsStore.GetWorkItem(workItemId);
                 workItem.Open();
 
-                fileList.ForEach(file => workItem.Attachments.Add(new Attachment(file)));
+                foreach (var file in fileList)
+                    workItem.Attachments.Add(new Attachment(file));
+
+                ValidateAndSaveWorkItem(workItem);
+            }
+            catch (Exception exception)
+            {
+                Logger.Error(exception.ToString());
+            }
+        }
+
+        public void AttachAndInlineFiles(int workItemId, IEnumerable<Tuple<string, IIncomingEmailAttachment>> fileList)
+        {
+            if (workItemId <= 0) return;
+
+            string fieldNameToUpdate = _config.WorkItemSettings.EmailBodyFieldName;
+
+            if (_config.EmailSettings.ConvertInlineAttachments && String.IsNullOrEmpty(fieldNameToUpdate))
+            {
+                Logger.WarnFormat("Inlining of images as attachments is not available because EmailBodyFieldName has not been set");
+                AttachFiles(workItemId, fileList.Select(f => f.Item1));
+                return;
+            }
+
+            try
+            {
+                var workItem = _tfsStore.GetWorkItem(workItemId);
+                workItem.PartialOpen();
+
+                var attachmentTrackers = fileList.Select(f => new
+                {
+                    WorkItemAttachmentIndex = workItem.Attachments.Add(new Attachment(f.Item1)),
+                    IsInline = f.Item2.IsInline,
+                    ContentId = f.Item2.ContentId
+                });
+
+                ValidateAndSaveWorkItem(workItem); // Save before proceeding to generate the URIs for the attachments
+
+                workItem.PartialOpen();
+
+                string html = workItem.Fields[fieldNameToUpdate].Value?.ToString();
+
+                const string pattern = @"(\<img.* src=\"")(cid:)(.*)(\"".*\>)";
+
+                html = Regex.Replace(html, pattern, match =>
+                {
+                    string contentId = match.Groups[3].ToString();
+
+                    var tracker = attachmentTrackers.FirstOrDefault(a => a.ContentId == contentId);
+                    if (tracker != null)
+                    {
+                        Logger.DebugFormat("Converting attached image to inline reference");
```

Comment:
There are two types are inlining that can occur. Base64 encoding defines the image content directly in the image src attribute. This is handled in the new class and the attachment is discarded since it has been "consumed" and is no longer needed

Base64 encoding only works for smaller images though. Quoting from this blog [post](http://johannblais.blogspot.com/2015/08/migrate-tfs-repro-steps-field-to.html), _"For an unknown reason, when the data in the src exceeds a certain limit (I don't know the actual value), the src attribute is blanked during the save. Images simply appear blank in the target field."_ So I picked a safe max size for inlining via base64 that worked in my test cases. For larger images, a traditional image reference is used, which has to be done after the attachment is created in TFS so that the attachment URI can be obtained. This is the code in TFSWorkItemManager that you're asking about. In this case, the new EWSIncomingAttachmentConverter class had captured information needs for this later step, i.e., in the FileAttachment type, the IsInline and ContentId properties.

In the latter case, I subsequently remove the attachment after linking the image to it because I had observed that TFS did not delete the file from the host system when the attachment was removed. At the time I thought it a nicety that the attachment list only contained "true" attachments -- void of images that are included in the item body. However, now that you bring this up I'm questioning whether this is a long-term safe assuming. What do you think?